examples/infra/mod.rs

```diff
@@ -54,8 +58,281 @@ use libp2p::{
+        Membership = MEMBERSHIP,
+        Networking = NETWORK,
+        Storage = MemoryStorage<TYPES, ValidatingLeaf<TYPES>>,
+    >,
+> where
+    <TYPES as NodeType>::StateType: TestableState,
+    <TYPES as NodeType>::BlockType: TestableBlock,
+    ValidatingLeaf<TYPES>: TestableLeaf,
+    HotShot<TYPES::ConsensusType, TYPES, NODE>: ViewRunner<TYPES, NODE>,
+    Self: Sync,
+{
+    /// Initializes networking, returns self
+    async fn initialize_networking(
+        config: NetworkConfig<TYPES::SignatureKey, TYPES::ElectionConfigType>,
+    ) -> Self;
+
+    /// Initializes the genesis state and HotShot instance; does not start HotShot consensus
+    async fn initialize_state_and_hotshot(&self) -> (TYPES::StateType, HotShotHandle<TYPES, NODE>) {
+        let genesis_block = TYPES::BlockType::genesis();
+        let initializer =
+            hotshot::HotShotInitializer::<TYPES, ValidatingLeaf<TYPES>>::from_genesis(
+                genesis_block,
+            )
+            .unwrap();
+        let config = self.get_config();
+        let (pub_key, secret_key) =
+            TYPES::SignatureKey::generated_from_seed_indexed(config.seed, config.node_index);
+        let known_nodes = config.config.known_nodes.clone();
+        let network = self.get_network();
+
+        // Since we do not currently pass the election config type in the NetworkConfig, this will always be the default election config
+        let election_config = config.config.election_config.clone().unwrap_or_else(|| {
+            NODE::Membership::default_election_config(config.config.total_nodes.get() as u64)
+        });
+
+        let hotshot = HotShot::init(
+            pub_key,
+            secret_key,
+            config.node_index,
+            config.config,
+            network,
+            MemoryStorage::new(),
+            MEMBERSHIP::create_election(known_nodes, election_config),
+            initializer,
+            NoMetrics::new(),
+        )
+        .await
+        .expect("Could not initialize hotshot");
+
+        let state = hotshot.storage().get_anchored_view().await.unwrap().state;
+        (state, hotshot)
+    }
+
+    /// Starts HotShot consensus, returns when consensus has finished
+    async fn run_hotshot(&self, mut hotshot: HotShotHandle<TYPES, NODE>) {
+        let NetworkConfig {
+            padding,
+            rounds,
+            transactions_per_round,
+            node_index,
+            config: HotShotConfig { total_nodes, .. },
+            ..
+        } = self.get_config();
+
+        let size = mem::size_of::<TYPES::Transaction>();
+        let adjusted_padding = if padding < size { 0 } else { padding - size };
+        let mut txns: VecDeque<TYPES::Transaction> = VecDeque::new();
+        let state = hotshot.get_state().await;
+
+        // This assumes that no node will be a leader more than 5x the expected number of times they should be the leader
+        // FIXME  is this a reasonable assumption when we start doing DA?
```

Comment:
I'm not clear about this. For now: super reasonable assumption. Long term: I think we're sampling from a stake distribution, which might result in a leader owning much of the stake. 

How is this assumption being used?